fairseq/search.py

```diff
@@ -75,11 +100,315 @@ def step(self, step: int, lprobs, scores: Optional[Tensor]):
         )
         scores_buf = top_prediction[0]
         indices_buf = top_prediction[1]
+        # Project back into relative indices and beams
         beams_buf = indices_buf // vocab_size
         indices_buf = indices_buf.fmod(vocab_size)
+
+        # At this point, beams_buf and indices_buf are single-dim and contain relative indices
         return scores_buf, indices_buf, beams_buf
 
 
+class LexicallyConstrainedBeamSearch(Search):
+    """Implements lexically constrained beam search by maintaining, for
+    each beam hypothesis, an object to track which of the constraints
+    have been met by that hypothesis.
+    """
```

Comment:
It might make sense to reference the paper(s) here.

Reply:
Will do.